docs/RFCS/20180411_alter_column_type.md

```diff
@@ -0,0 +1,303 @@
+completed, this command will block until it has.  It is not an error to
+run `COMPLETE` before the conversion job has finished; it is simply a
+signal to indicate that the SQL consumers are prepared to handle the new
+column type.
+
+## Key constraints
+
+Columns that are the target of a primary or foreign key constraint will
+not be supported under this RFC.  It could be possible to support
+foreign keys in the future by cascading the type change, but this would
+greatly increase the amount of coordination required to effect the total
+schema change.
+
+# Reference-level explanation
+
+## Process
+
+Conceptually, a column type change can be thought of as creating a
+shadow column of the new type, which captures writes to the original,
+source, column while a backfill process is run.  Once the backfill is
+complete, the shadow column is swapped for the original, which is then
+retired.  The shadow column will be anonymous and generally invisible to
+the user until it is swapped in.
+
+This list describes the general type-change case, not all of these steps
+are necessarily executed for trivial or validated changes.
+
+* `ALTER TABLE ... COLUMN ... TYPE` is run, resulting in a new shadow
+  column being created in `DELETE_ONLY` mode.  The shadow column has a
+  new column id and a reference to the source column
+  * If there are indexes that refer to the column, additional
+    shadow-indexes are created.
+* The shadow column and indexes are upgraded to `WRITE_ONLY` mode.
+* Nodes that write to the source column will synchronously apply the
+  requested type change and duplicate the write to the shadow column
+  and indexes.
+* A backfill process is distributed across all nodes in the cluster
+  which have data for the table.  This backfill will populate the shadow
+  column and shadow indexes, inserting data with the source MVCC
+  timestamp. The backfilled sstables will be inserted into the LSM such
+  that any shadowed write will have priority.
+* The following steps will only occur if the change is in auto-complete
+  mode or when the user issues the `COMPLETE` command.
+* Once the backfill is complete, the `WRITE_ONLY` restriction is removed
+  from the shadow column and shadow indexes.  Any code that resolves a
+  column or index name to an id should be prepared to look for a shadow
+  that has been activated.
+* The table metadata is updated, changing the column name and indexes
+  to point at the shadow data, making the original column or indexes
+  non-live.
+* The old column and indexes are retired in the same manner as dropping
+  a column, allowing any active readers to complete their transaction.
+
+To summarize:
+* Column and index writers should be prepared to duplicate their writes
+  when a writable shadow exists for their target column or index.
+* Code that resolves a column or index name for reading should be
+  prepared to look for an activated shadow.
+* It must be possible to insert bulk, backfilled data into the LSM such
```

Comment:
I'm not sure this is a thing we want to do because it would actually change history if someone was doing a read AS OF SYSTEM TIME.